[1.4] town NPC profile system extension + NPC name improvements

ThomasThePencil opened the pull request:
# IN WHICH
the town NPC profile system partially introduced by vanilla is expanded upon

## why does this exist?
to make the town NPC profile system more accessible to modders --- this also has the effect of allowin' modders to actually change the Princess' sprites without havin' to use such measures as an unstable bout of reflection. since it was related, I also felt it was appropriate to make some NPC namin' system changes that I feel have been needed for a very long time

## changelog
- added town NPC profiles for all town NPCs not covered normally (the single unifyin' factor between all of them is that none of them have alternate textures when wearin' a party hat)
- added the new `(ModNPC/GlobalNPC).SetTownNPCProfile` hook, actin' as a way for modders to not only add their own town NPC profiles, but also remove or replace those that already exist
- repurposed `ModNPC.TownNPCName` into `ModNPC.SetNPCNameList`, which allows returnin' a list of all the names a given modded NPC can have
- added `GlobalNPC.ModifyNPCNameList`, allowin' people to change the possible names of existin' NPCs easily
- added `(ModNPC/GlobalNPC).ModifyTypeName`, finally allowin' people to change the type names of new and existin' NPCs (and do so dynamically, for that matter!) without havin' to engage in an obscenely large amount of tomfoolery
- as a result of needin' them to make the new changes work, added a few new methods to `LanguageManager` involvin' localization categories

## portin' notes (what you need to know)
`ModNPC.TownNPCName()` ---> `ModNPC.SetNPCNameList`, now returns a `List<string>` instead of a `string`

ThomasThePencil committed:
added all vanilla town NPCs to the town NPC profile system

mod extensibility comin' soon

Rijam commented:
This might be beyond the scope of this PR, but could `AltTextures` be reimplemented for Town NPCs? `AltTextures` is the texture that the Town NPCs switch to when a party is active.

ThomasThePencil commented:
these are still a thing, and every town NPC included in vanilla's profile list uses them with a few exceptions. I recommend lookin' into vanilla's `LegacyNPCProfile`s to see how it's done

ThomasThePencil committed:
SetTownNPCProfile is the new hook in town

woohoo

SheepishShepherd commented:
Im not too familiar with how things are turning out but I am working on a mod that could potentially benefit from this. Would methods like the `GetTextureNPCShouldUse` or `GetNameForVariant` be overridden/modified to add new conditional textures/names. If so, how would multiple mods be able to edit such things and not override each other.

ThomasThePencil commented:
this really just finishes off the system for vanilla town NPCs, and adds the ability to add new town NPC profiles to the list and remove existin' ones --- this does allow replacement, of course, but I can see how that'd become problematic with multiple mods in play

I'm not sure what exactly can be done to alleviate this issue without a notably larger-scale refactor, but I'd be willin' to look into any alternatives/additions proposed nonetheless

ThomasThePencil committed:
get real (ExampleMod changes)

ThomasThePencil committed:
actual EM fix lol

Solxanich commented on patches/tModLoader/Terraria/GameContent/TownNPCProfiles.cs.patch:
```diff
@@ -0,0 +1,60 @@
+ 			{ 227, new Profiles.LegacyNPCProfile("Images/TownNPCs/Painter", 17) },
++			{ NPCID.PartyGirl, new Profiles.AlternateLegacyNPCProfile("Images/NPC_208", NPCHeadID.PartyGirl) },
+ 			{ 229, new Profiles.LegacyNPCProfile("Images/TownNPCs/Pirate", 19) },
++			{ 663, new Profiles.LegacyNPCProfile("Images/TownNPCs/Princess", 45) },
```
Is there a reason why this line was moved? Accidental change?

ThomasThePencil replied:
I had wanted to organize them alphabetically but now that I think about it that falls apart with the town pets so I might just move Princess back to her original spot

Solxanich commented on patches/tModLoader/Terraria/GameContent/TownNPCProfiles.cs.patch:
```diff
@@ -0,0 +1,60 @@
+ 			{ 638, new Profiles.VariantNPCProfile("Images/TownNPCs/Dog", "Dog", DogHeadIDs, "Labrador", "PitBull", "Beagle", "Corgi", "Dalmation", "Husky") },
+ 			{ 656, new Profiles.VariantNPCProfile("Images/TownNPCs/Bunny", "Bunny", BunnyHeadIDs, "White", "Angora", "Dutch", "Flemish", "Lop", "Silver") }		};
++
++		public Dictionary<int, ITownNPCProfile> TownNPCProfileList;
```
This probably should be localized instead of held as part of the instance. 
I'm not seeing any calls to it outside of the one method

ThomasThePencil replied:
I had originally done that and then had second thoughts about it. good to see I was right the first time :HDfailure:

Solxanich commented on patches/tModLoader/Terraria/GameContent/TownNPCProfiles.cs.patch:
```diff
@@ -0,0 +1,60 @@
+ 			43,
+ 			44
+ 		};
++		// does Old Man count? should he? need to reference vanilla or ask someone who's done so already     -thomas
```
Is this Todo sorted, or is this still an open action?

ThomasThePencil replied:
I have no idea, but he didn't show up there even though Clothier (his non-cursed variant) did, so I'll assume Old Man isn't included in this and strike this todo accordingly

Solxanich reviewed (commented):
Some minor changes.

Overall looks reasonable, I do see breaking change? in terms of including the Texture override & how the name gets assigned to the npc. 
We shouldn't have 2 ways of accomplishing the same outcome, so best if we stick to just the new option being available. 
Thoughts?

ThomasThePencil committed:
moved Princess back to her original location

I hate this orderin' but whatever
also localized the town NPC profile list for tidiness and nixed the Old Man todo since from what I'm aware he's not actually counted in the way that normal town NPCs are (he wasn't included even though Traveling Merchant and Skeleton Merchant, two other not-really-town NPCs with the same setup, did get included)

ThomasThePencil commented:
> Overall looks reasonable, I do see breaking change? in terms of including the Texture override
as an aside, the Texture override is included here so that the NPC texture autoload system doesn't prevent EM from loadin'. vanilla has a good way of namin' these

ThomasThePencil committed:
A LOT OF SHIT

e
- getNewNPCName refactored immensely to take the NPC itself rather than the type alone
- repurposed `ModNPC.TownNPCName` into `ModNPC.SetNPCNameList`
- added `GlobalNPC.ModifyNPCNameList
- added a few new methods to LanguageManager because I needed them to make this work lmao

ThomasThePencil committed:
docu fixes lo

ThomasThePencil committed:
type name shenanigans shall finally get real

by relatively popular demand, and because I have several use cases, the (ModNPC/GlobalNPC).ModifyTypeName hook has been added

ThomasThePencil committed:
general cleanup

- changed the way firstNPCName is killed so that it takes up less patch space
- tidied up the massive switch case used for vanilla town NPC names
- removed the non-raw version of GetLocalizedEntriesInCategory and made the raw version named that

Solxanich reviewed (approved):
Looks generally liveable. 
A TODO will need to be kept to de-hardcode the switch case, but that's all internal/minimal modder impacting. 

I'm good for now, noting the above TODO.
